enumerables.rb

```diff
@@ -0,0 +1,126 @@
+module Enumerable
+  def my_each
+    return to_enum unless block_given?
+
+    length.times { |x| yield self[x] }
+    self
+  end
+
+  def my_each_with_index
+    return to_enum unless block_given?
+
+    length.times { |x| yield self[x], x }
+    self
+  end
+
+  def my_select
+    return to_enum unless block_given?
+
+    ret_arr = []
+    my_each do |x|
+      next unless yield(x)
+
+      ret_arr << x
+    end
+    ret_arr
+  end
+
+  # rubocop:disable Metrics/PerceivedComplexity,Metrics/CyclomaticComplexity
+
+  def my_all?(pattern = nil)
+    return false if include?(nil) || include?(false)
+
+    my_each do |x|
+      if block_given?
+        return false unless yield x
+      elsif pattern.class == Regexp
+        return false unless pattern =~ x
+      elsif pattern.class == Class
+        return false unless x.class == pattern
+      else
+        return true
+      end
+    end
+    true
+  end
+
+  def my_any?(pattern = nil)
+    return true if include?(nil)
+    return true if include?(false)
+
+    my_each do |x|
+      if block_given?
+        return true if yield x
+      elsif pattern.class == Regexp
+        return true if pattern =~ x
+      elsif pattern.class == Class
+        return true if x.class == pattern
+      else
+        return false
+      end
+    end
+    false
+  end
+
+  def my_none?(pattern = nil)
+    return true if include?(true)
+
+    my_each do |x|
+      if block_given?
+        return false if yield x
+      elsif pattern.class == Regexp
+        return false if pattern =~ x
+      elsif pattern.class == Class
+        return false if x.class == pattern
+      else
+        return true
+      end
+    end
+    true
+  end
+
+  # rubocop:enable Metrics/PerceivedComplexity,Metrics/CyclomaticComplexity
+
+  def my_count(item = nil)
+    return length if item.nil? && !block_given?
+
+    count = 0
+    my_each do |x|
+      if block_given?
+        count += 1 if yield x
+      elsif x == item
+        count += 1
+      end
+    end
+
+    count
+  end
+
+  def my_map(proc = nil)
+    return to_enum unless block_given?
+
+    ret_arr = []
+    my_each { |x| ret_arr << (!proc.nil? ? proc.call(x) : yield(x)) }
+    ret_arr
+  end
+
+  def my_inject(*args)
+    dummy = dup
+    unless block_given?
+
+      sum = (args.length.positive? ? args.shift : sum = dummy.shift)
+
+      dummy.my_each { |x| sum = sum.send(args[0].to_s, x) }
+      return sum
+    end
+
+    sum = (args.length.positive? ? args[0] : dummy.shift)
+
+    dummy.my_each { |x| sum = yield sum, x }
+    sum
+  end
```

Comment:
`my_inject` identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- combines all elements of enum by applying a binary operation, specified by a block
```ruby
[1, 2, 3].my_inject { |memo, num| memo + num } #should return 6
[1, 2, 3].my_inject(1) { |memo, num| memo + num } #should return 7
```
- when a symbol is specified combines each element of the collection by applying the symbol as a named method
```ruby
[1, 2, 3].my_inject(:+)  #should return 6
[1, 2, 3].my_inject(1, :+)  #should return 7
```

Reply:
Done! Thank you for letting me know. I added some extra conditions to make it return the correct values.